src/unix/index_unix.ml

```diff
@@ -254,6 +254,34 @@ module IO : Index.IO = struct
       let _ = Unix.fstat t.raw.fd in
       true
     with Unix.Unix_error (Unix.EBADF, _, _) -> false
+
+  type fd = Unix.file_descr
+
+  let _lock op f =
+    mkdir (Filename.dirname f);
+    let fd = Unix.openfile f [ Unix.O_CREAT; Unix.O_RDWR ] 0o600
+    and pid = string_of_int (Unix.getpid ()) in
+    let pid_len = String.length pid in
+    try
+      Unix.lockf fd op 0;
+      if Unix.single_write_substring fd pid 0 pid_len <> pid_len then (
+        Unix.close fd;
+        failwith "Unable to write PID to lock file" )
+      else Some fd
+    with
+    | Unix.Unix_error (Unix.EAGAIN, _, _) ->
+        Unix.close fd;
+        None
+    | e ->
+        Unix.close fd;
+        raise e
+
+  let lock path =
+    match _lock Unix.F_LOCK path with
```

Comment:
maybe use `F_RLOCK` to have a proper exception?

Reply:
Then both tests cannot run, one of them raises the exception. 

Do I remove the concurrent test ?

Reply:
yes, we should have a deterministic way to test this. For instance, use `Unix.exec` with a process trying to open that file and check the exit code.

For an end-user perspective, the role of that lock file is to give a feedback quickly that the API is used wrongly (e.g. there *must* be only one writer process). Blocking will not really help here :-)

Reply:
For the moment I remove the concurrent test altogether, so that you can merge this if you want. I'll add the concurrent tests once I figured out how to use `Unix.exec`.